pkg/gadgets/trace/exec/tracer/tracer_test.go

```diff
@@ -0,0 +1,208 @@
+	}
+	if tracer == nil {
+		t.Fatal("Returned tracer was nil")
+	}
+
+	// Check that a double stop doesn't cause issues
+	tracer.Stop()
+	tracer.Stop()
+}
+
+func createTracer(
+	t *testing.T, config *tracer.Config, callback func(types.Event),
+) *tracer.Tracer {
+	t.Helper()
+
+	tracer, err := tracer.NewTracer(config, nil, callback)
+	if err != nil {
+		t.Fatalf("Error creating tracer: %s", err)
+	}
+	return tracer
+}
+
+func TestTracer(t *testing.T) {
+	utilstest.RequireRoot(t)
+
+	manyArgs := []string{}
+	// 19 is DEFAULT_MAXARGS - 1 (-1 because args[0] is on the first position).
+	for i := 0; i < 19; i++ {
+		manyArgs = append(manyArgs, "/dev/null")
+	}
+
+	// Function to generate an event used most of the times.
+	// Returns pid of executed process.
+	generateEvent := func() (int, error) {
+		cmd := exec.Command("/bin/cat", "/dev/null")
+		if err := cmd.Run(); err != nil {
+			return 0, fmt.Errorf("running command: %w", err)
+		}
+
+		return cmd.Process.Pid, nil
+	}
+
+	tests := map[string]utilstest.Test[types.Event, int]{
+		"captures_all_events_without_configured_filters": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				return createTracer(t, &tracer.Config{}, eventCallback)
+			},
+			GenerateEvent: generateEvent,
+			ValidateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.Info, ret *int) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					Pid:       uint32(*ret),
+					Ppid:      uint32(info.Pid),
+					UID:       uint32(info.UID),
+					MountNsID: info.MountNsID,
+					Retval:    0,
+					Comm:      "cat",
+					Args:      []string{"/bin/cat", "/dev/null"},
+				}
+			}),
+		},
+		"filtering_no_event_should_be_captured": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				config := &tracer.Config{
+					MountnsMap: utilstest.CreateMntNsFilterMap(t, 0),
+				}
+				return createTracer(t, config, eventCallback)
+			},
+			GenerateEvent: generateEvent,
+			ValidateEvent: utilstest.ExpectNoEvent[types.Event, int],
+		},
+		"filtering_one_event_should_be_captured": {
+			GetTracer: func(info *utilstest.Info, eventCallback func(event types.Event)) utilstest.Tracer {
+				config := &tracer.Config{
+					MountnsMap: utilstest.CreateMntNsFilterMap(t, info.MountNsID),
+				}
+				return createTracer(t, config, eventCallback)
+			},
+			GenerateEvent: generateEvent,
+			ValidateEvent: utilstest.ExpectOneEvent(func(info *utilstest.Info, ret *int) *types.Event {
+				return &types.Event{
+					Event: eventtypes.Event{
+						Type: eventtypes.NORMAL,
+					},
+					Pid:       uint32(*ret),
+					Ppid:      uint32(info.Pid),
+					UID:       uint32(info.UID),
+					MountNsID: info.MountNsID,
+					Retval:    0,
+					Comm:      "cat",
+					Args:      []string{"/bin/cat", "/dev/null"},
+				}
+			}),
+		},
+		"event_has_correct_UID": {
```

Comment:
Use of word `corrrect` here is ambiguous IMO and the event definition here is largely dynamically generated. I wonder if it wouldn't be better to get slice of events here and assert by hand that all of them has expected PID used for configuring runner. I think that would be much better showing the intention of the test.

It would also much better highlight expected properties of events, as right now you use DeepEqual and one must define entire event for test to pass, which puts a lot of unrelated details into the test definition, which hides the essence of what exactly is under test. Of course, this could be workaround by having some helper function generating "base" event using `info` too.

Reply:
I was writing a long message indicating it was difficult to implement and that we'll lost the advantage of having  `ExpectOneEvent`, but then I realized that with `ValidateEvent` it's so easy to implement. I agree, I changed it to only test the UID.